lib/quicly.c

```diff
@@ -3324,25 +3321,21 @@ static int handle_ack_frame(quicly_conn_t *conn, size_t epoch, quicly_ack_frame_
     /* OnPacketAcked */
     uint32_t latest_rtt = UINT32_MAX, ack_delay = 0;
     if (largest_newly_acked.packet_number == frame->largest_acknowledged) {
-        int64_t t = now - largest_newly_acked.sent_at;
-        if (0 <= t && t < 100000) { /* ignore RTT above 100 seconds */
-            latest_rtt = (uint32_t)t;
-            uint64_t ack_delay_microsecs = frame->ack_delay << conn->super.peer.transport_params.ack_delay_exponent;
-            ack_delay = (uint32_t)((ack_delay_microsecs * 2 + 1000) / 2000);
-        }
+        uint64_t ack_delay_microsecs = frame->ack_delay << conn->super.peer.transport_params.ack_delay_exponent;
+        ack_delay = (uint32_t)((ack_delay_microsecs * 2 + 1000) / 2000);
+        /* Ignore samples where ack_delay is larger than max_ack_delay.  Note: This is a departure from the recovery
+         * draft which uses this RTT sample, but limiting the ack_delay to min(ack_delay, max_ack_delay). This departure
+         * allows us to not ignore RTT samples where the largest acked is in fact not ack-eliciting, since we simply ignore
+         * the sample where ack_delay is too large for such a packet. */
+        if (ack_delay > conn->super.peer.transport_params.max_ack_delay)
+                latest_rtt = (uint32_t)(now - largest_newly_acked.sent_at);
     }
 
-    quicly_loss_on_ack_received(
-        &conn->egress.loss, frame->largest_acknowledged, latest_rtt, ack_delay,
-        0 /* this relies on the fact that we do not (yet) retransmit ACKs and therefore latest_rtt becoming UINT32_MAX */);
-
-    if (smallest_newly_acked != UINT64_MAX)
-        quicly_loss_on_newly_acked(&conn->egress.loss);
+    quicly_loss_on_ack_received(&conn->egress.loss, frame->largest_acknowledged, latest_rtt, ack_delay);
 
-    /* OnPacketAckedCC */
-    /* TODO (jri): this function should be called for every packet newly acked. (kazuho) I do not think so;
-     * quicly_loss_on_packet_acked is NOT OnPacketAcked */
+    /* OnPacketAcked and OnPacketAckedCC */
     if (bytes_acked > 0) {
+        quicly_loss_on_newly_acked(&conn->egress.loss);
```

Comment:
Shouldn't we call `quicly_loss_on_newly_acked` when we receive an ACK for any new PN?

Assuming that is the case, the condition for calling this particular function should be `largest_newly_acked.packet_number != UINT64_MAX`. `bytes_acked > 0` is a test that checks if any ACK-eliciting packet is newly acknowledged before it was deemed lost. I think that is fine for the other calls that we make in this block.

FWIW, I think there might be an error in the [pseudo-code of recovery spec](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#rfc.appendix.B.5). `OnPacketAckedCC` substracts `acked_packet.size` from the number of bytes inflight. I believe that should not be done if the packet has already declared lost.

Reply:
On the code: Yes, good point, and it was an oversight. I've changed it now.
On the pseudo-code: Good catch! I'll file an issue.